### PR TITLE
Added unit tests for AwsIamAuthenticationPlugin

### PR DESCRIPTION
**Pull request by alecc-bq**
…AwsIamAuthenticationPlugin

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Summary:

Added gradle dependencies for aws sdk and mockito and created unit tests for AwsIamAuthenticationPlugin.java class. 

### Description:

- Created test to catch exception if wrong type of authentication request is made
- Created test to retrieve a valid token from the cache to provide authentication
- Created test to generate a token given an expired token in the cache
- Created test to generate a token given no token is found in the cache

### CODE REVIEW

File: pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
-import software.amazon.awssdk.regions.Region;
-import software.amazon.awssdk.services.rds.RdsUtilities;
-
 import org.postgresql.PGProperty;
 import org.postgresql.plugin.AuthenticationPlugin;
 import org.postgresql.plugin.AuthenticationRequestType;
 import org.postgresql.util.GT;
 import org.postgresql.util.PSQLException;
 
+import org.checkerframework.checker.nullness.qual.Nullable;
```
Comment (brunos-bq) [inline pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java:13]: Is this the correct order of imports? Can you check if the project has guidelines about import order?

File: pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java [hunk 2]
```diff
@@ -156,7 +155,7 @@ private Region getRdsRegion(final String hostname) throws PSQLException {
     return regionOptional.get();
   }
 
-  private static class TokenInfo {
+  public static class TokenInfo {
```
Comment (brunos-bq) [inline pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java]: do you need this to be public? Or just some public accessors?

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import java.lang.reflect.Modifier;
+import java.time.Instant;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class AwsIamAuthenticationPluginTest {
+
+  @Test
+  public void testGetPasswordWithWrongType() throws PSQLException {
+    Properties properties = new Properties();
```
Comment (brunos-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: The first 4 lines are repeated in every test.

Can you check on how to use @Before and @BeforeEach and remove those from the test code?

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    properties.setProperty(PGProperty.USER.getName(), "testUser");
```
Comment (brunos-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: Hard coded values should be in some global variables

File: pgjdbc/build.gradle.kts [hunk 5]
```diff
@@ -56,7 +56,7 @@ dependencies {
 
     compileOnly("software.amazon.awssdk:rds:2.17.189")
     compileOnly("software.amazon.awssdk:auth:2.17.189")
-
+    implementation("software.amazon.awssdk:rds:2.17.210")
```
Comment (karenc-bq) [inline pgjdbc/build.gradle.kts]: I think this should be a test implementation

File: pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java [hunk 6]
```diff
@@ -97,7 +96,7 @@ private String getCacheKey(
     return String.format("%s:%s:%d:%s", region, hostname, port, user);
   }
 
-  private String generateAuthenticationToken(
+  protected String generateAuthenticationToken(
```
Comment (karenc-bq) [inline pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java]: package private

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
+package software.aws.rds.jdbc.postgresql.auth;
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: Don't forget the license

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:5]:
```suggestion
```

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    AwsIamAuthenticationPlugin spyPlugin = Mockito.spy(new AwsIamAuthenticationPlugin(properties));
+    when(spyPlugin.generateAuthenticationToken("testUser", "test.user.us-east-2.rds.amazonaws.com", 5342, Region.US_EAST_2)).thenReturn("generatedToken");
+    char[] actualResult = spyPlugin.getPassword(AuthenticationRequestType.CLEARTEXT_PASSWORD);
+    assertArrayEquals("generatedToken".toCharArray(), actualResult);
+  }
+
+}
+
+
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]:
```suggestion
```

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    plugin.getClass().getDeclaredField("tokenCache").setAccessible(true);
+    Field tokenField = plugin.getClass().getDeclaredField("tokenCache");
+    tokenField.setAccessible(true);
+    Field modifiersField = Field.class.getDeclaredField("modifiers");
+    modifiersField.setAccessible(true);
+    modifiersField.setInt(tokenField, tokenField.getModifiers() & ~Modifier.FINAL);
+    ConcurrentHashMap<String, AwsIamAuthenticationPlugin.TokenInfo> value = new ConcurrentHashMap<>();
+    value.clear();
+    value.put("us-east-2:test.user.us-east-2.rds.amazonaws.com:5342:testUser", new AwsIamAuthenticationPlugin.TokenInfo("testToken", Instant.now().plusMillis(300000)));
+    tokenField.set(plugin, value);
```
Comment (sergiyvamz) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: Why do you need to override cache object with on new one? You can use a cache object the has been initialized by plugin.

File: pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java (hunk 2 above)
Comment (alecc-bq) [reply pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java]: Will make package private.

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  public void testGetPasswordWithWrongType() throws PSQLException {
+    Properties properties = new Properties();
+    properties.setProperty(PGProperty.USER.getName(), "testUser");
+    properties.setProperty(PGProperty.PG_HOST.getName(), "test.user.us-east-2.rds.amazonaws.com");
+    AwsIamAuthenticationPlugin plugin = new AwsIamAuthenticationPlugin(properties);
+    assertThrows(PSQLException.class, () -> plugin.getPassword(AuthenticationRequestType.MD5_PASSWORD));
+  }
+
+  @Test
+  public void testGetPasswordValidTokenInCache() throws PSQLException, NoSuchFieldException,
```
Comment (sergiyvamz) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: A general note. Blank lines in the code increase readability.

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+      IllegalAccessException {
+    Properties properties = new Properties();
+    properties.setProperty(PGProperty.USER.getName(), "testUser");
+    properties.setProperty(PGProperty.PG_HOST.getName(), "test.user.us-east-2.rds.amazonaws.com");
+    AwsIamAuthenticationPlugin plugin = new AwsIamAuthenticationPlugin(properties);
```
Comment (sergiyvamz) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: plugin -> targetPlugin ?

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 10 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: addressed in most recent changes

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 12 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: addressed in most recent changes

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  private static final String GENERATED_TOKEN = "generatedToken";
+  private static final String CACHE_KEY = "us-east-2:test.user.us-east-2.rds.amazonaws.com:5342:testUser";
+
+  @BeforeEach
+  public void initTargetPlugin() throws PSQLException {
+    properties = new Properties();
+    properties.setProperty(PGProperty.USER.getName(), TEST_USER);
+    properties.setProperty(PGProperty.PG_HOST.getName(), TEST_HOST);
+    targetPlugin = new AwsIamAuthenticationPlugin(properties);
+  }
```
Comment (sergiyvamz) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:40]: AwsIamAuthenticationPlugin.tokenCache.clear()

We have to clear cache before each test

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2022, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+package software.aws.rds.jdbc.postgresql.auth;
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:6]: There should be a line between the license and the package path

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+    assertArrayEquals(GENERATED_TOKEN.toCharArray(), actualResult);
+  }
+
+  @Test
+  public void testGetPasswordGenerateToken() throws PSQLException {
+    AwsIamAuthenticationPlugin.tokenCache.clear();
+    AwsIamAuthenticationPlugin spyPlugin = Mockito.spy(new AwsIamAuthenticationPlugin(properties));
+    when(spyPlugin.generateAuthenticationToken(TEST_USER, TEST_HOST, 5342, Region.US_EAST_2)).thenReturn(GENERATED_TOKEN);
+    char[] actualResult = spyPlugin.getPassword(AuthenticationRequestType.CLEARTEXT_PASSWORD);
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: final

Comment (karenc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: same for the others

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  @Test
+  public void testGetPasswordValidTokenInCache() throws PSQLException {
+    AwsIamAuthenticationPlugin.tokenCache.put(CACHE_KEY, new AwsIamAuthenticationPlugin.TokenInfo(TEST_TOKEN, Instant.now().plusMillis(300000)));
+    char[] actualResult = targetPlugin.getPassword(AuthenticationRequestType.CLEARTEXT_PASSWORD);
+
+    assertArrayEquals(TEST_TOKEN.toCharArray(), actualResult);
+  }
+
+  @Test
+  public void testGetPasswordExpiredTokenInCache() throws PSQLException {
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:56]: We should also test getPassword with empty cache

Comment (karenc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:56]: Nevermind it is tested in `testGetPasswordGenerateToken`

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  @Test
+  public void testGetPasswordGenerateToken() throws PSQLException {
+    AwsIamAuthenticationPlugin.tokenCache.clear();
+    AwsIamAuthenticationPlugin spyPlugin = Mockito.spy(new AwsIamAuthenticationPlugin(properties));
+    when(spyPlugin.generateAuthenticationToken(TEST_USER, TEST_HOST, 5342, Region.US_EAST_2)).thenReturn(GENERATED_TOKEN);
+    char[] actualResult = spyPlugin.getPassword(AuthenticationRequestType.CLEARTEXT_PASSWORD);
+
+    assertArrayEquals(GENERATED_TOKEN.toCharArray(), actualResult);
```
Comment (karenc-bq) [inline pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:72]: We should also check the content of the token cache here.

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 16 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:56]: I am renaming the test to testGetPasswordEmptyCache()

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 14 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:6]: :+1:

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 13 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:40]: :+1:

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 15 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: not quite sure what exactly this means?

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 17 above)
Comment (alecc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java:72]: working on this :+1:

File: pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java (hunk 15 above)
Comment (karenc-bq) [reply pgjdbc/src/test/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPluginTest.java]: These variables can have the final keyword since they are never changed.
`final char[] actualResult = spyPlugin.getPassword(AuthenticationRequestType.CLEARTEXT_PASSWORD);`

File: pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java [hunk 18]
```diff
@@ -1,16 +1,15 @@
 package software.aws.rds.jdbc.postgresql.auth;
```
Comment (karenc-bq) [inline pgjdbc/src/main/java/software/aws/rds/jdbc/postgresql/auth/AwsIamAuthenticationPlugin.java:6]: Please add a license for this file as well.
